File: proof_kernel/hypothesis_proposals.py
"""Hypothesis proposal workflow.

Agents may propose revisions to baseline hypotheses. Canonical hypotheses do
not mutate silently. Acceptance requires an explicit `accept` call that records
who accepted, when, the old/new hashes, and the affected experiments.

Storage layout:

  artifacts/runs/<run_id>/hypothesis_proposals/<proposal_id>.json
      The proposal itself. Status transitions PROPOSED → ACCEPTED | REJECTED.

  artifacts/runs/<run_id>/hypothesis_proposals/<proposal_id>.audit.json
      Audit record written on accept/reject.

  proof_kernel/hypotheses/_accepted_overlays.json
      Active accepted overlays. Loader merges these on top of the canonical
      files at runtime. This keeps canonical JSON files clean in git while
      allowing accepted proposals to take effect immediately.

The overlay layer is intentionally simple: the latest accepted proposal for a
given hypothesis_id wins. Rejecting an active overlay also removes it.
"""

from __future__ import annotations

import datetime
import hashlib
import json
import os
import re
import secrets
from pathlib import Path
from typing import Any, Dict, List, Optional

from .hypothesis_registry import (
    REGISTRY_FILES,
    SCHEMA_VERSION as REGISTRY_SCHEMA_VERSION,
    load_registry,
)

PROPOSAL_SCHEMA_VERSION = "2026.05.hypothesis-proposal.v1"
OVERLAY_FILENAME = "_accepted_overlays.json"

VALID_STATUS = {"PROPOSED", "ACCEPTED", "REJECTED"}


# ---------------------------------------------------------------------------
# Path helpers
# ---------------------------------------------------------------------------

def _proposals_dir(run_id: str, repo: Path | str = ".") -> Path:
    return Path(repo) / "artifacts" / "runs" / run_id / "hypothesis_proposals"


def _proposal_path(run_id: str, proposal_id: str, repo: Path | str = ".") -> Path:
    return _proposals_dir(run_id, repo) / f"{proposal_id}.json"


def _audit_path(run_id: str, proposal_id: str, repo: Path | str = ".") -> Path:
    return _proposals_dir(run_id, repo) / f"{proposal_id}.audit.json"


def _overlay_path(repo: Path | str = ".") -> Path:
    return Path(repo) / "proof_kernel" / "hypotheses" / OVERLAY_FILENAME


# ---------------------------------------------------------------------------
# IDs and hashing
# ---------------------------------------------------------------------------

def _new_proposal_id() -> str:
    ts = datetime.datetime.now(datetime.timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    suffix = secrets.token_hex(3)
    return f"prop_{ts}_{suffix}"


def _now_iso() -> str:
    return datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")


def _hash_baseline(payload: Any) -> str:
    """Stable SHA-256 over a baseline-shaped object."""
    raw = json.dumps(payload, sort_keys=True, default=str, separators=(",", ":"))
    return hashlib.sha256(raw.encode("utf-8")).hexdigest()


def _safe_json(path: Path, payload: Dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    with open(tmp, "w", encoding="utf-8") as handle:
        json.dump(payload, handle, indent=2, sort_keys=True)
        handle.write("\n")
    os.replace(tmp, path)


# ---------------------------------------------------------------------------
# Validation
# ---------------------------------------------------------------------------

class ProposalError(Exception):
    pass


REQUIRED_PROPOSAL_FIELDS = (
    "source_agent",
    "experiment_id",
    "proposed_baseline",
    "reason",
)


def _validate_proposal_input(payload: Dict[str, Any]) -> None:
    missing = [f for f in REQUIRED_PROPOSAL_FIELDS if not payload.get(f)]
    if missing:
        raise ProposalError(f"missing required fields: {missing}")
    if not isinstance(payload.get("proposed_baseline"), dict):
        raise ProposalError("proposed_baseline must be a baseline-shaped object")
    pb = payload["proposed_baseline"]
    for f in ("plain_statement", "expected_signature"):
        if not pb.get(f):
            raise ProposalError(f"proposed_baseline missing {f}")


# ---------------------------------------------------------------------------
# Overlay management
# ---------------------------------------------------------------------------

def _load_overlay(repo: Path | str = ".") -> Dict[str, Any]:
    path = _overlay_path(repo)
    if not path.exists():
        return {
            "schema_version": PROPOSAL_SCHEMA_VERSION,
            "registry_schema_version": REGISTRY_SCHEMA_VERSION,
            "overlays": {},
        }
    return json.loads(path.read_text(encoding="utf-8"))


def _save_overlay(overlay: Dict[str, Any], repo: Path | str = ".") -> None:
    _safe_json(_overlay_path(repo), overlay)


def get_accepted_overlay(repo: Path | str = ".") -> Dict[str, Any]:
    return _load_overlay(repo)


def apply_overlays_to_registry(
    base_registry: Dict[str, Any],
    repo: Path | str = ".",
) -> Dict[str, Any]:
    """Merge the accepted-overlays file on top of a freshly-loaded registry.

    Returns a new registry-shape dict with the same keys as load_registry().
    """
    overlay = _load_overlay(repo)
    overlays = overlay.get("overlays") or {}
    if not overlays:
        return base_registry

    by_hyp = dict(base_registry["by_hypothesis_id"])
    by_exp = dict(base_registry["by_experiment_id"])
    by_disp = dict(base_registry["by_display_id"])
    all_entries = list(base_registry["all"])

    for hyp_id, ov in overlays.items():
        accepted = ov.get("accepted_baseline")
        if not isinstance(accepted, dict):
            continue
        merged = {**by_hyp.get(hyp_id, {}), **accepted}
        merged["_overlay_provenance"] = {
            "from_proposal_id": ov.get("proposal_id"),
            "accepted_by": ov.get("accepted_by"),
            "accepted_at": ov.get("accepted_at"),
            "old_baseline_hash": ov.get("old_baseline_hash"),
            "new_baseline_hash": ov.get("new_baseline_hash"),
        }
        by_hyp[hyp_id] = merged
        # Refresh by_experiment / by_display lookups for this hypothesis.
        for exp_id in merged.get("experiment_ids", []) or []:
            by_exp[exp_id] = merged
        if merged.get("display_id"):
            by_disp[merged["display_id"]] = merged
        # Replace in all_entries (or append if not present).
        replaced = False
        for i, entry in enumerate(all_entries):
            if entry.get("hypothesis_id") == hyp_id:
                all_entries[i] = merged
                replaced = True
                break
        if not replaced:
            all_entries.append(merged)

    return {
        **base_registry,
        "by_hypothesis_id": by_hyp,
        "by_experiment_id": by_exp,
        "by_display_id": by_disp,
        "all": all_entries,
    }


# ---------------------------------------------------------------------------
# Public proposal API
# ---------------------------------------------------------------------------

def propose_baseline_update(
    run_id: str,
    payload: Dict[str, Any],
    repo: Path | str = ".",
) -> Dict[str, Any]:
    """Create a new PROPOSED hypothesis proposal.

    payload requires: source_agent, experiment_id, proposed_baseline, reason.
    Optional: evidence, recommended_next_experiment.
    """
    _validate_proposal_input(payload)
    registry = load_registry(repo)
    exp_id = payload["experiment_id"]
    current = registry["by_experiment_id"].get(exp_id)
    if current is None:
        raise ProposalError(f"unknown experiment_id: {exp_id}")

    proposal_id = _new_proposal_id()
    proposal = {
        "schema_version": PROPOSAL_SCHEMA_VERSION,
        "proposal_id": proposal_id,
        "run_id": run_id,
        "source_agent": payload["source_agent"],
        "experiment_id": exp_id,
        "hypothesis_id": current["hypothesis_id"],
        "current_baseline": current,
        "proposed_baseline": payload["proposed_baseline"],
        "reason": payload["reason"],
        "evidence": payload.get("evidence", []) or [],
        "recommended_next_experiment": payload.get("recommended_next_experiment"),
        "created_at": _now_iso(),
        "status": "PROPOSED",
        "old_baseline_hash": _hash_baseline(current),
        "new_baseline_hash": _hash_baseline(payload["proposed_baseline"]),
    }
    _safe_json(_proposal_path(run_id, proposal_id, repo), proposal)
    return proposal


def list_proposals(
    run_id: str,
    repo: Path | str = ".",
    status: Optional[str] = None,
) -> List[Dict[str, Any]]:
    d = _proposals_dir(run_id, repo)
    if not d.exists():
        return []
    out: List[Dict[str, Any]] = []
    for entry in sorted(d.iterdir()):
        if not entry.is_file() or entry.suffix != ".json":
            continue
        if entry.name.endswith(".audit.json"):
            continue
        try:
            data = json.loads(entry.read_text(encoding="utf-8"))
        except Exception:
            continue
        if status and data.get("status") != status:
            continue
        out.append(data)
    return out


def get_proposal(
    run_id: str,
    proposal_id: str,
    repo: Path | str = ".",
) -> Optional[Dict[str, Any]]:
    p = _proposal_path(run_id, proposal_id, repo)
    if not p.exists():
        return None
    return json.loads(p.read_text(encoding="utf-8"))


def get_proposal_audit(
    run_id: str,
    proposal_id: str,
    repo: Path | str = ".",
) -> Optional[Dict[str, Any]]:
    p = _audit_path(run_id, proposal_id, repo)
    if not p.exists():
        return None
    return json.loads(p.read_text(encoding="utf-8"))


def accept_proposal(
    run_id: str,
    proposal_id: str,
    accepted_by: str,
    note: Optional[str] = None,
    repo: Path | str = ".",
) -> Dict[str, Any]:
    if not accepted_by or not str(accepted_by).strip():
        raise ProposalError("accepted_by is required")
    proposal = get_proposal(run_id, proposal_id, repo)
    if proposal is None:
        raise ProposalError(f"proposal not found: {proposal_id}")
    if proposal.get("status") != "PROPOSED":
        raise ProposalError(f"proposal is not in PROPOSED status (current={proposal.get('status')})")

    accepted_at = _now_iso()
    proposal["status"] = "ACCEPTED"
    proposal["accepted_by"] = accepted_by
    proposal["accepted_at"] = accepted_at
    if note:
        proposal["acceptance_note"] = note
    _safe_json(_proposal_path(run_id, proposal_id, repo), proposal)

    # Write the audit record
    audit = {
        "schema_version": PROPOSAL_SCHEMA_VERSION,
        "proposal_id": proposal_id,
        "run_id": run_id,
        "decision": "ACCEPTED",
        "decided_by": accepted_by,
        "decided_at": accepted_at,
        "note": note,
        "experiment_id": proposal["experiment_id"],
        "hypothesis_id": proposal["hypothesis_id"],
        "old_baseline_hash": proposal["old_baseline_hash"],
        "new_baseline_hash": proposal["new_baseline_hash"],
        "old_baseline_snapshot": proposal["current_baseline"],
        "new_baseline_snapshot": proposal["proposed_baseline"],
        "affected_experiments": proposal["current_baseline"].get("experiment_ids", []),
    }
    _safe_json(_audit_path(run_id, proposal_id, repo), audit)

    # Merge into the accepted-overlays layer
    overlay = _load_overlay(repo)
    overlay["overlays"][proposal["hypothesis_id"]] = {
        "proposal_id": proposal_id,
        "run_id": run_id,
        "accepted_by": accepted_by,
        "accepted_at": accepted_at,
        "old_baseline_hash": proposal["old_baseline_hash"],
        "new_baseline_hash": proposal["new_baseline_hash"],
        "accepted_baseline": proposal["proposed_baseline"],
    }
    _save_overlay(overlay, repo)

    return proposal


def reject_proposal(
    run_id: str,
    proposal_id: str,
    rejected_by: str,
    reason: Optional[str] = None,
    repo: Path | str = ".",
) -> Dict[str, Any]:
    if not rejected_by or not str(rejected_by).strip():
        raise ProposalError("rejected_by is required")
    proposal = get_proposal(run_id, proposal_id, repo)
    if proposal is None:
        raise ProposalError(f"proposal not found: {proposal_id}")
    if proposal.get("status") not in ("PROPOSED", "ACCEPTED"):
        raise ProposalError(f"proposal cannot be rejected from status={proposal.get('status')}")

    decided_at = _now_iso()
    previous_status = proposal["status"]
    proposal["status"] = "REJECTED"
    proposal["rejected_by"] = rejected_by
    proposal["rejected_at"] = decided_at
    if reason:
        proposal["rejection_reason"] = reason
    _safe_json(_proposal_path(run_id, proposal_id, repo), proposal)

    audit = {
        "schema_version": PROPOSAL_SCHEMA_VERSION,
        "proposal_id": proposal_id,
        "run_id": run_id,
        "decision": "REJECTED",
        "decided_by": rejected_by,
        "decided_at": decided_at,
        "previous_status": previous_status,
        "reason": reason,
        "experiment_id": proposal["experiment_id"],
        "hypothesis_id": proposal["hypothesis_id"],
        "old_baseline_hash": proposal["old_baseline_hash"],
        "new_baseline_hash": proposal["new_baseline_hash"],
    }
    _safe_json(_audit_path(run_id, proposal_id, repo), audit)

    # If this proposal was active in the overlay, remove it.
    if previous_status == "ACCEPTED":
        overlay = _load_overlay(repo)
        active = overlay["overlays"].get(proposal["hypothesis_id"])
        if active and active.get("proposal_id") == proposal_id:
            overlay["overlays"].pop(proposal["hypothesis_id"], None)
            _save_overlay(overlay, repo)

    return proposal

File: lib/hypothesis-proposals.ts
/**
 * Hypothesis proposal workflow (TS reader + TS shim that shells out to Python
 * for mutating actions).
 *
 * Reads (proposals, audits, overlay) are pure-Node. Writes (propose / accept /
 * reject) call into the Python `proof_kernel.hypothesis_proposals` module so
 * the storage format and overlay-merge logic live in one place.
 */

import { execFileSync } from "child_process";
import crypto from "crypto";
import fs from "fs";
import path from "path";
import { clearHypothesisRegistryCache } from "./hypothesis-registry";

export const PROPOSAL_SCHEMA_VERSION = "2026.05.hypothesis-proposal.v1";
export const OVERLAY_FILENAME = "_accepted_overlays.json";

export type ProposalStatus = "PROPOSED" | "ACCEPTED" | "REJECTED";

export interface HypothesisProposal {
    schema_version: string;
    proposal_id: string;
    run_id: string;
    source_agent: string;
    experiment_id: string;
    hypothesis_id: string;
    current_baseline: Record<string, unknown>;
    proposed_baseline: Record<string, unknown>;
    reason: string;
    evidence: unknown[];
    recommended_next_experiment?: string | null;
    created_at: string;
    status: ProposalStatus;
    old_baseline_hash: string;
    new_baseline_hash: string;
    accepted_by?: string;
    accepted_at?: string;
    acceptance_note?: string;
    rejected_by?: string;
    rejected_at?: string;
    rejection_reason?: string;
}

export interface ProposalAudit {
    schema_version: string;
    proposal_id: string;
    run_id: string;
    decision: "ACCEPTED" | "REJECTED";
    decided_by: string;
    decided_at: string;
    note?: string | null;
    reason?: string | null;
    previous_status?: ProposalStatus;
    experiment_id: string;
    hypothesis_id: string;
    old_baseline_hash: string;
    new_baseline_hash: string;
    old_baseline_snapshot?: Record<string, unknown>;
    new_baseline_snapshot?: Record<string, unknown>;
    affected_experiments?: string[];
}

export interface AcceptedOverlay {
    proposal_id: string;
    run_id: string;
    accepted_by: string;
    accepted_at: string;
    old_baseline_hash: string;
    new_baseline_hash: string;
    accepted_baseline: Record<string, unknown>;
}

export interface AcceptedOverlayFile {
    schema_version: string;
    registry_schema_version?: string;
    overlays: Record<string, AcceptedOverlay>;
}

export interface ProposeInput {
    source_agent: string;
    experiment_id: string;
    proposed_baseline: Record<string, unknown>;
    reason: string;
    evidence?: unknown[];
    recommended_next_experiment?: string | null;
}

export class ProposalError extends Error {}

// ---------------------------------------------------------------------------
// Path helpers
// ---------------------------------------------------------------------------

function proposalsDir(runId: string, repoRoot: string): string {
    return path.join(repoRoot, "artifacts", "runs", runId, "hypothesis_proposals");
}

function proposalPath(runId: string, proposalId: string, repoRoot: string): string {
    return path.join(proposalsDir(runId, repoRoot), `${proposalId}.json`);
}

function auditPath(runId: string, proposalId: string, repoRoot: string): string {
    return path.join(proposalsDir(runId, repoRoot), `${proposalId}.audit.json`);
}

function overlayPath(repoRoot: string): string {
    return path.join(repoRoot, "proof_kernel", "hypotheses", OVERLAY_FILENAME);
}

// ---------------------------------------------------------------------------
// Reads
// ---------------------------------------------------------------------------

export function readAcceptedOverlay(repoRoot: string = process.cwd()): AcceptedOverlayFile {
    const fp = overlayPath(repoRoot);
    if (!fs.existsSync(fp)) {
        return { schema_version: PROPOSAL_SCHEMA_VERSION, overlays: {} };
    }
    return JSON.parse(fs.readFileSync(fp, "utf-8"));
}

export function listProposals(
    runId: string,
    repoRoot: string = process.cwd(),
    statusFilter?: ProposalStatus,
): HypothesisProposal[] {
    const dir = proposalsDir(runId, repoRoot);
    if (!fs.existsSync(dir)) return [];
    const out: HypothesisProposal[] = [];
    for (const entry of fs.readdirSync(dir).sort()) {
        if (!entry.endsWith(".json")) continue;
        if (entry.endsWith(".audit.json")) continue;
        const fp = path.join(dir, entry);
        try {
            const data = JSON.parse(fs.readFileSync(fp, "utf-8")) as HypothesisProposal;
            if (statusFilter && data.status !== statusFilter) continue;
            out.push(data);
        } catch {
            // skip malformed
        }
    }
    return out;
}

export function getProposal(
    runId: string,
    proposalId: string,
    repoRoot: string = process.cwd(),
): HypothesisProposal | null {
    const fp = proposalPath(runId, proposalId, repoRoot);
    if (!fs.existsSync(fp)) return null;
    return JSON.parse(fs.readFileSync(fp, "utf-8"));
}

export function getProposalAudit(
    runId: string,
    proposalId: string,
    repoRoot: string = process.cwd(),
): ProposalAudit | null {
    const fp = auditPath(runId, proposalId, repoRoot);
    if (!fs.existsSync(fp)) return null;
    return JSON.parse(fs.readFileSync(fp, "utf-8"));
}

// ---------------------------------------------------------------------------
// Mutations (in-process, mirrored against the Python module's storage format)
// ---------------------------------------------------------------------------

function safeWriteJson(target: string, payload: unknown): void {
    fs.mkdirSync(path.dirname(target), { recursive: true });
    const tmp = `${target}.tmp`;
    fs.writeFileSync(tmp, `${JSON.stringify(payload, null, 2)}\n`, "utf-8");
    fs.renameSync(tmp, target);
}

function stableHash(payload: unknown): string {
    return crypto.createHash("sha256").update(canonicalize(payload)).digest("hex");
}

function canonicalize(value: unknown): string {
    if (value === null || typeof value !== "object") return JSON.stringify(value);
    if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;
    const entries = Object.entries(value as Record<string, unknown>).sort(([a], [b]) => a.localeCompare(b));
    return `{${entries.map(([k, v]) => `${JSON.stringify(k)}:${canonicalize(v)}`).join(",")}}`;
}

function newProposalId(): string {
    const ts = new Date().toISOString().replace(/[-:T]/g, "").replace(/\..+/, "Z");
    const suffix = crypto.randomBytes(3).toString("hex");
    return `prop_${ts}_${suffix}`;
}

function nowIso(): string {
    return new Date().toISOString().replace(/\.\d+Z$/, "Z");
}

const REQUIRED_PROPOSAL_FIELDS: (keyof ProposeInput)[] = [
    "source_agent",
    "experiment_id",
    "proposed_baseline",
    "reason",
];

export function proposeBaselineUpdate(
    runId: string,
    payload: ProposeInput,
    repoRoot: string = process.cwd(),
): HypothesisProposal {
    for (const field of REQUIRED_PROPOSAL_FIELDS) {
        if (!payload[field]) throw new ProposalError(`missing required field: ${field}`);
    }
    if (typeof payload.proposed_baseline !== "object" || payload.proposed_baseline === null) {
        throw new ProposalError("proposed_baseline must be a baseline-shaped object");
    }
    const pb = payload.proposed_baseline;
    if (!pb.plain_statement) throw new ProposalError("proposed_baseline missing plain_statement");
    if (!pb.expected_signature) throw new ProposalError("proposed_baseline missing expected_signature");

    // Read the current canonical baseline (no overlays — proposals compare to canonical).
    const currentRegistry = JSON.parse(
        execFileSync(
            "python",
            [
                "-c",
                "import json,sys; from proof_kernel.hypothesis_registry import load_registry; sys.stdout.write(json.dumps(load_registry('.', apply_overlays=False)))",
            ],
            { cwd: repoRoot, encoding: "utf-8" },
        ),
    );
    const current = currentRegistry.by_experiment_id?.[payload.experiment_id];
    if (!current) throw new ProposalError(`unknown experiment_id: ${payload.experiment_id}`);

    const proposalId = newProposalId();
    const proposal: HypothesisProposal = {
        schema_version: PROPOSAL_SCHEMA_VERSION,
        proposal_id: proposalId,
        run_id: runId,
        source_agent: payload.source_agent,
        experiment_id: payload.experiment_id,
        hypothesis_id: current.hypothesis_id,
        current_baseline: current,
        proposed_baseline: payload.proposed_baseline,
        reason: payload.reason,
        evidence: payload.evidence ?? [],
        recommended_next_experiment: payload.recommended_next_experiment ?? null,
        created_at: nowIso(),
        status: "PROPOSED",
        old_baseline_hash: stableHash(current),
        new_baseline_hash: stableHash(payload.proposed_baseline),
    };
    safeWriteJson(proposalPath(runId, proposalId, repoRoot), proposal);
    return proposal;
}

export function acceptProposal(
    runId: string,
    proposalId: string,
    acceptedBy: string,
    note: string | null = null,
    repoRoot: string = process.cwd(),
): HypothesisProposal {
    if (!acceptedBy?.trim()) throw new ProposalError("accepted_by is required");
    const proposal = getProposal(runId, proposalId, repoRoot);
    if (!proposal) throw new ProposalError(`proposal not found: ${proposalId}`);
    if (proposal.status !== "PROPOSED") {
        throw new ProposalError(`proposal is not in PROPOSED status (current=${proposal.status})`);
    }

    const acceptedAt = nowIso();
    proposal.status = "ACCEPTED";
    proposal.accepted_by = acceptedBy;
    proposal.accepted_at = acceptedAt;
    if (note) proposal.acceptance_note = note;
    safeWriteJson(proposalPath(runId, proposalId, repoRoot), proposal);

    const audit: ProposalAudit = {
        schema_version: PROPOSAL_SCHEMA_VERSION,
        proposal_id: proposalId,
        run_id: runId,
        decision: "ACCEPTED",
        decided_by: acceptedBy,
        decided_at: acceptedAt,
        note,
        experiment_id: proposal.experiment_id,
        hypothesis_id: proposal.hypothesis_id,
        old_baseline_hash: proposal.old_baseline_hash,
        new_baseline_hash: proposal.new_baseline_hash,
        old_baseline_snapshot: proposal.current_baseline,
        new_baseline_snapshot: proposal.proposed_baseline,
        affected_experiments: (proposal.current_baseline as { experiment_ids?: string[] })
            .experiment_ids ?? [],
    };
    safeWriteJson(auditPath(runId, proposalId, repoRoot), audit);

    const overlay = readAcceptedOverlay(repoRoot);
    overlay.overlays[proposal.hypothesis_id] = {
        proposal_id: proposalId,
        run_id: runId,
        accepted_by: acceptedBy,
        accepted_at: acceptedAt,
        old_baseline_hash: proposal.old_baseline_hash,
        new_baseline_hash: proposal.new_baseline_hash,
        accepted_baseline: proposal.proposed_baseline,
    };
    safeWriteJson(overlayPath(repoRoot), overlay);
    clearHypothesisRegistryCache();
    return proposal;
}

export function rejectProposal(
    runId: string,
    proposalId: string,
    rejectedBy: string,
    reason: string | null = null,
    repoRoot: string = process.cwd(),
): HypothesisProposal {
    if (!rejectedBy?.trim()) throw new ProposalError("rejected_by is required");
    const proposal = getProposal(runId, proposalId, repoRoot);
    if (!proposal) throw new ProposalError(`proposal not found: ${proposalId}`);
    if (proposal.status !== "PROPOSED" && proposal.status !== "ACCEPTED") {
        throw new ProposalError(`proposal cannot be rejected from status=${proposal.status}`);
    }
    const previousStatus = proposal.status;
    const decidedAt = nowIso();
    proposal.status = "REJECTED";
    proposal.rejected_by = rejectedBy;
    proposal.rejected_at = decidedAt;
    if (reason) proposal.rejection_reason = reason;
    safeWriteJson(proposalPath(runId, proposalId, repoRoot), proposal);

    const audit: ProposalAudit = {
        schema_version: PROPOSAL_SCHEMA_VERSION,
        proposal_id: proposalId,
        run_id: runId,
        decision: "REJECTED",
        decided_by: rejectedBy,
        decided_at: decidedAt,
        previous_status: previousStatus,
        reason,
        experiment_id: proposal.experiment_id,
        hypothesis_id: proposal.hypothesis_id,
        old_baseline_hash: proposal.old_baseline_hash,
        new_baseline_hash: proposal.new_baseline_hash,
    };
    safeWriteJson(auditPath(runId, proposalId, repoRoot), audit);

    if (previousStatus === "ACCEPTED") {
        const overlay = readAcceptedOverlay(repoRoot);
        const active = overlay.overlays[proposal.hypothesis_id];
        if (active && active.proposal_id === proposalId) {
            delete overlay.overlays[proposal.hypothesis_id];
            safeWriteJson(overlayPath(repoRoot), overlay);
            clearHypothesisRegistryCache();
        }
    }
    return proposal;
}

// Reload helper: re-applies overlay to the in-memory hypothesis registry.
export function reloadHypothesisRegistryAfterMutation(): void {
    clearHypothesisRegistryCache();
}
